`gravatar`- Add API Key initialization to the SDK

hamorillo opened the pull request:
Closes #156 

### Description

Adding the method to initialize the SDK with the API Key. At the moment, that API Key is required by the new REST API when you want to get a full profile. Without the API  Key, some fields are not returned by the backend because you need to be authenticated.

Also adding in the DemoApp an expandable section that allows to review the profile raw data:

<img width="250" alt="image" src="https://github.com/Automattic/Gravatar-SDK-Android/assets/6974554/22574a03-ae7c-4647-bef2-7a561c7f93d6">  <img width="250" alt="image" src="https://github.com/Automattic/Gravatar-SDK-Android/assets/6974554/6c48c9ff-f514-4111-9574-eaea371c9393">

### Testing Steps

- Add your ApiKey in `local.properties` as follows:
```
demo-app.gravatar.api.key=YOUR_API_KEY_HERE
```
- Launch the DemoApp
- Go to the `Profile` tab
- Get any profile
- Expand the `Raw Profile`
- Verify it contains fields only present within authenticated requests like: `registrationDate`, `lastProfileEdit`, `contactInfo` etc.

- You can also verify that you don't receive those fields when the ApiKey is removed from `local.properties`.

hamorillo committed:
Add Gravatar SDK initialisation method

Clients should use the API key generated in the Gravatar Developer portal to get complete information from the Gravatar SDK.

When you don't use the API Key you'll get the basic information. That info is the minimum needed to show the Profile components in the UI module.

We are providing a method to initialise the SDK with the API Key. After that, all the requests will use that key.

hamorillo committed:
Enabling DemoApp to use Gravatar API Key

Load gravatar.api.key from local.properties to initialize the SDK.

hamorillo committed:
Add "Raw Profile" in DemoApp

As the answer returned by the backend when using the API Key contains more information, it would be nice to have a simple way to see the Profile content.

hamorillo commented:
I'll keep the PR as a draft because we need some backend changes. But feel free if you want to review the code or run the app without the ApiKey (with the API Key will crash until the changes are done). @mlumeau

hamorillo commented:
Backend changes have been deployed, so the PR is ready to review.

hamorillo committed:
Update to last API specification

The authentication bearer has been defined globally, so when we generate the client, it's not present as a parameter in each method. In our case, in the "getProfileById".

We are creating an interceptor to add the bearer in each request.

mlumeau committed:
Improved interceptor readability and maintainability

mlumeau commented on gravatar/src/main/java/com/gravatar/di/container/GravatarSdkContainer.kt:
```diff
@@ -53,7 +56,7 @@ internal class GravatarSdkContainer private constructor() {
     fun getGravatarApiV3Service(okHttpClient: OkHttpClient? = null): GravatarApiService {
         return getRetrofitApiV3Builder().apply {
-            okHttpClient?.let { client(it) }
+            client(okHttpClient ?: OkHttpClient().newBuilder().addInterceptor(AuthenticationInterceptor()).build())
```
We're not adding our interceptor to any dev-supplied HTTP client? Not necessarily saying we should, but since we made the AuthenticationInterceptor `internal`, devs will have to re-write it in order to support an API key...

Here's what this change would look like:

```suggestion
            client((okHttpClient ?: OkHttpClient()).newBuilder().addInterceptor(AuthenticationInterceptor()).build())
```

hamorillo replied:
It's a good point. Tbh, I had the same doubt. I've committed the suggested change.

mlumeau commented on gravatar/src/main/java/com/gravatar/services/AuthenticationInterceptor.kt:
```diff
@@ -0,0 +1,20 @@
+import okhttp3.Response
+import com.gravatar.di.container.GravatarSdkContainer.Companion.instance as GravatarSdkDI
+
+internal class AuthenticationInterceptor : Interceptor {
```
I pushed a quick modification in order to improve readability and facilitate adding other headers if needed in the future. Tell me what you think.

mlumeau reviewed (commented):
I just have a question regarding third party HTTP clients. Once we agree I'll approve the PR.

hamorillo committed:
Add AuthenticationInterceptor to any OkHttpClient when using API v3

We should add the authenticator interceptor to both OkHttpClient. It doesn't matter if it is a parameter or internally created by the SDK.

mlumeau reviewed (approved):
LGTM 🚀
